### ISSUE TITLE
✨ [Feature] Add CSRF token middleware like Laravel

### ISSUE THREAD

**Issue by DevHaoZi** [goravel/goravel#218]
### Before feedback (在反馈之前)

- [X] There are no features that I want to submit in Issues (当前 Issues 中没有我将要提交的新功能)

### Describe Feature (描述功能)

Add CSRF token middleware like Laravel.

**Comment by NiteshSingh17** [comment]
Hi @hwbrzzl 

I’ve been looking into this issue and was exploring a potential solution. My idea is to pass a struct (with a String() method implemented) as a shared variable using the `GetShared` function with the key `csrf_token`:
[https://github.com/goravel/framework/blob/ff6dace8b2c6c8dd2e132a8cc58a7496cb1724d0/http/view.go#L39](https://github.com/goravel/framework/blob/ff6dace8b2c6c8dd2e132a8cc58a7496cb1724d0/http/view.go?utm_source=chatgpt.com#L39)

However, when the user accesses this variable in their template, I need to set a session value on the request. For that, I’ll need access to the request object.

Would it be acceptable to pass the request object to the `GetShared` function (from the `goravel/gin `library), or would that break any design constraints or architectural patterns?

Thanks!

**Comment by krishankumar01** [comment]
@NiteshSingh17, Goravel sessions already include a `_token` attribute (accessible via the `Token()` method) that was originally added for CSRF support. We’re not using it right now, but the token is regenerated on every session regenerate and can be tweaked, you can consider using `_token`.

https://github.com/goravel/framework/blob/ff6dace8b2c6c8dd2e132a8cc58a7496cb1724d0/contracts/session/session.go#L54

**Comment by NiteshSingh17** [comment]
I have a couple of questions regarding the CSRF token implementation that I’d appreciate your help.

1. Regarding how we pass csrf_token to templates:
Are we expecting the user to manually pass the csrf_token from the session in every response, like in the following example?
```

facades.Route().Get("/", func(ctx http.Context) http.Response {
    // Get the CSRF token from session
    csrfToken := ctx.Session().Token()

    return ctx.Response().View().Make("welcome.tmpl", map[string]any{
        "version": support.Version,
        "csrf_token": csrfToken, // Pass the token to your view
    })
})

```
Or are we planning to automatically inject `csrf_token` into all views as a globally shared variable like larvel? I’m assuming the latter is preferred for consistency and to reduce boilerplate, something like:
```

type CSRFType struct {}

func (c CSRFType) String() string {
    return "csrf_token"
}

func (r *View) GetShared() map[string]any {
    shared := make(map[string]any)
    r.shared.Range(func(key, value any) bool {
        shared[key.(string)] = value
        return true
    })

    shared["csrf_token"] = CSRFType{}
    return shared
}

```
2. If we go with auto-population, how can I access `Session.Token()` within the context of the View.GetShared() method (or any similar place where we populate shared variables)?

Since I'm still getting familiar with the codebase, I’m not sure what’s the recommended way to access session data or the request context in this part of the flow. Any guidance or pointers on how that’s typically handled here would be really helpful.

**Comment by krishankumar01** [comment]
Yeah, I agree, the user shouldn't have to pass `csrf_token` manually. It should just be there.

Since `framework/session/middleware.StartSession` already adds the session to the context, the [`Gin`](https://github.com/goravel/gin) and [`Fiber`](https://github.com/goravel/fiber) packages should be able to grab the token from there and inject it into the view data automatically. For APIs, we can just send it back in a cookie for clients like `Axios` to use.

The best place to start looking would be in the respective `view` and `context_request` implementations for Gin and Fiber drivers, as well as in the `framework/session` package.

**Comment by NiteshSingh17** [comment]
I've been exploring the `Gin` repository and noticed that the `Session()` method is available on the `http.ContextRequest`. However, when rendering views, the `ContextResponse` is used, which does not have access to the request object and therefore cannot access session data directly.

Looking at the view creation logic:

View Creator:(https://github.com/goravel/gin/blob/d5b89f1c1055a124ab5140f017bcaf436a2a4faf/view.go?utm_source=chatgpt.com#L12)

ContextResponse:(https://github.com/goravel/gin/blob/d5b89f1c1055a124ab5140f017bcaf436a2a4faf/context_response.go?utm_source=chatgpt.com#L22)

Given this structure, what would be the recommended way to access session data within the context of `ContextResponse`?

Would it be appropriate to pass the request (or just the session) into `NewContextResponse`, or is there a preferred pattern within the framework for accessing request related data during view rendering?

**Comment by hwbrzzl** [comment]
We may implement a custom template method: `{{ csrf_token .ctx }}` first, then verify the csrf token in the csrf token middleware.
